LV_USE_CHART is hardcoded to 0 in generated lv_conf.h with no supported override mechanism

### The problem

Component: LVGL
ESPHome version: 2026.4.1
Description:
Since the LVGL v8 → v9 migration in 2026.4.0, the LVGL chart widget (lv_chart) is completely inaccessible from user lambdas. ESPHome generates lv_conf.h with #define LV_USE_CHART 0, which compiles out the entire chart widget. There is currently no supported way to override this:

build_flags: ["-DLV_USE_CHART=1"] is overridden by the generated header (produces a "redefined" warning, the header wins)
build_unflags cannot remove a #define inside a generated header file
esphome: includes: injects too late — after LVGL's #if LV_USE_CHART guards have already compiled the widget out
-DLV_CONF_PATH=... redirecting to a custom config file is itself redefined by ESPHome's own CMake configuration (visible as "LV_CONF_PATH" redefined warnings in the build log)

What we'd like:
Either one of these would solve it completely:

A YAML option under lvgl: to enable additional widgets — e.g. widgets: [chart] or extra_widgets: [chart] — that causes ESPHome to set LV_USE_CHART 1 in the generated config
Documentation of a supported override mechanism if one exists

Why it matters:
The LVGL chart widget is the natural fit for displaying sensor history on an embedded display — temperature, CO₂, humidity over time. Before 2026.4.0 it was achievable via lambdas with the v8 API. The v9 migration removed that path with no replacement.
Workaround: Pin to ESPHome 2026.3.3.

### Which version of ESPHome has the issue?

2026.4.0

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

ESP32

### Component causing the issue

LVGL

### YAML Config

```yaml
substitutions:
  device_name: esphome-chart-test

esphome:
  name: ${device_name}
  friendly_name: Chart Test
  platformio_options:
    board_build.flash_mode: dio
    build_flags:
      - "-D LV_USE_CHART=1"
  on_boot:
    priority: -100
    then:
      - lambda: |-
          // Attempt to create an LVGL chart widget via lambda.
          // This is currently the only available mechanism as ESPHome does
          // not expose the chart widget natively in its LVGL YAML schema.
          auto *chart = lv_chart_create(id(chart_placeholder));
          lv_obj_set_size(chart, lv_pct(100), lv_pct(100));
          lv_chart_set_point_count(chart, 96);
          lv_chart_set_range(chart, LV_CHART_AXIS_PRIMARY_Y, 0, 100);
          id(graph_obj) = chart;
          id(graph_series) = lv_chart_add_series(
            chart,
            lv_color_make(0, 255, 0),
            LV_CHART_AXIS_PRIMARY_Y
          );

esp32:
  board: lolin_s2_mini
  variant: ESP32S2
  framework:
    type: esp-idf

logger:
api:
ota:
  - platform: esphome

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

psram:

spi:
  clk_pin: GPIO01
  mosi_pin: GPIO02
  miso_pin: GPIO16

# Minimal display to host LVGL
display:
  - platform: ili9xxx
    model: ILI9341
    cs_pin: GPIO03
    dc_pin: GPIO04
    reset_pin: GPIO06
    #data_rate: 10MHz
    transform:
      #mirror_x: true
      mirror_y: true
    #show_test_card: true
    color_order: bgr # CHANGE THIS if your screen has colors mapped wrong
    #color_palette: 8BIT
    invert_colors: false
    dimensions:
      height: 320
      width: 240
    auto_clear_enabled: false
    update_interval: never

# Dummy sensor to feed the chart
sensor:
  - platform: template
    id: test_sensor
    name: "Test Sensor"
    update_interval: 15min
    lambda: |-
      return 42.0;

globals:
  - id: graph_obj
    type: lv_obj_t *
    restore_value: no
  - id: graph_series
    type: void *
    restore_value: no

lvgl:
  id: lvgl_component
  color_depth: 16
  bg_color: 0
  pages:
    - id: main_page
      widgets:
        - obj:
            id: chart_placeholder
            align: CENTER
            width: 200
            height: 150
            bg_color: 0
            border_width: 0
            pad_all: 0

interval:
  - interval: 15min
    then:
      - lambda: |-
          if (id(graph_obj) != nullptr && id(graph_series) != nullptr) {
            lv_chart_set_next_value(
              id(graph_obj),
              (lv_chart_series_t*)id(graph_series),
              (int)id(test_sensor).state
            );
            lv_obj_invalidate(id(graph_obj));
          }
```

### Anything in the logs that might be useful for us?

```txt
In file included from managed_components/lvgl__lvgl/src/lv_conf_internal.h:68,
                 from managed_components/lvgl__lvgl/src/lv_init.h:16,
                 from managed_components/lvgl__lvgl/lvgl.h:21,
                 from src/esphome/components/lvgl/lvgl_proxy.h:13,
                 from src/esphome/core/color.h:8,
                 from src/esphome/components/display/display.h:8,
                 from src/esphome/components/display/display.cpp:1:
/data/build/esphome-chart-test/src/lv_conf.h:62:9: warning: "LV_USE_CHART" redefined
   62 | #define LV_USE_CHART 0
      |         ^~~~~~~~~~~~
<command-line>: note: this is the location of the previous definition
In file included from managed_components/lvgl__lvgl/src/lv_conf_internal.h:68,
                 from managed_components/lvgl__lvgl/src/lv_init.h:16,
                 from managed_components/lvgl__lvgl/lvgl.h:21,
                 from src/esphome/components/lvgl/lvgl_proxy.h:13,
                 from src/esphome/core/color.h:8,
                 from src/esphome/components/display/display.h:8,
                 from src/esphome/components/display/display_buffer.h:6,
                 from src/esphome/components/display/display_buffer.cpp:1:
/data/build/esphome-chart-test/src/lv_conf.h:62:9: warning: "LV_USE_CHART" redefined
   62 | #define LV_USE_CHART 0
      |         ^~~~~~~~~~~~
<command-line>: note: this is the location of the previous definition
```

### Additional information

With the help of claude I tried several workarounds but did not help.
In short, it was working fine in in prior LVGL version, producing very nice customisable charts. 
I can provide a full YAML, but its almost 3K lines. The example above compiles fine in 2026.3.3. But errors out in 2026.4.0-1

--- a/esphome/components/lvgl/__init__.py
+++ b/esphome/components/lvgl/__init__.py
@@ -155,13 +155,11 @@ def generate_lv_conf_h():
     build_flags = (
         CORE.config[CONF_ESPHOME].get(CONF_PLATFORMIO_OPTIONS).get("build_flags", [])
     )
-    # Extract define names from build flags like '-DLV_USE_CHART=1' or '-DLV_USE_CHART'
-    define_pattern = r'-D([A-Z_][A-Z0-9_]*)(?:=[^\s\'"\]]*)?'
+    # Extract define names from build flags like '-DLV_USE_CHART=1', '-D LV_USE_CHART',
+    # or multiple defines in one string.
+    define_pattern = r'-D\s*([A-Z_][A-Z0-9_]*)(?:=[^\s\'"\]]*)?'
     defines_from_flags = {
-        m.group(1)
-        for flag in build_flags
-        for m in [re.search(define_pattern, flag)]
-        if m
+        m.group(1) for flag in build_flags for m in re.finditer(define_pattern, flag)
     }
 
     # Get the defines that are actually used based on the config,
@@ -170,7 +168,7 @@ def generate_lv_conf_h():
     if clashes:
         LOGGER.warning(
             "Some defines are set both by ESPHome build flags and by LVGL configuration which may lead to unexpected behavior: %s",
-            list(clashes),
+            sorted(list(clashes)),
         )
     unused_defines = all_defines - lv_defines.keys() - defines_from_flags
 

--- a/esphome/components/lvgl/__init__.py
+++ b/esphome/components/lvgl/__init__.py
@@ -1,6 +1,7 @@
 import importlib
 from pathlib import Path
 import pkgutil
+import re
 
 from esphome.automation import Trigger, build_automation, validate_automation
 import esphome.codegen as cg
@@ -30,12 +31,14 @@ import esphome.config_validation as cv
 from esphome.const import (
     CONF_AUTO_CLEAR_ENABLED,
     CONF_BUFFER_SIZE,
+    CONF_ESPHOME,
     CONF_GROUP,
     CONF_ID,
     CONF_LAMBDA,
     CONF_LOG_LEVEL,
     CONF_ON_IDLE,
     CONF_PAGES,
+    CONF_PLATFORMIO_OPTIONS,
     CONF_ROTATION,
     CONF_TIMEOUT,
     CONF_TRIGGER_ID,
@@ -51,6 +54,7 @@ from . import defines as df, lv_validation as lvalid, widgets
 from .automation import layers_to_code, lvgl_update
 from .defines import (
     CONF_ALIGN_TO_LAMBDA_ID,
+    LOGGER,
     get_focused_widgets,
     get_lv_images_used,
     get_refreshed_widgets,
@@ -148,9 +152,28 @@ def generate_lv_conf_h():
     all_defines = set(
         df.LV_DEFINES + tuple(f"LV_USE_{w.upper()}" for w in WIDGET_TYPES)
     )
-    # Get the defines that are actually used based on the config
+    build_flags = (
+        CORE.config[CONF_ESPHOME].get(CONF_PLATFORMIO_OPTIONS).get("build_flags", [])
+    )
+    # Extract define names from build flags like '-DLV_USE_CHART=1' or '-DLV_USE_CHART'
+    define_pattern = r'-D([A-Z_][A-Z0-9_]*)(?:=[^\s\'"\]]*)?'
+    defines_from_flags = {
+        m.group(1)
+        for flag in build_flags
+        for m in [re.search(define_pattern, flag)]
+        if m
+    }
+
+    # Get the defines that are actually used based on the config,
     lv_defines = df.get_defines()
-    unused_defines = all_defines - set(lv_defines)
+    clashes = defines_from_flags & lv_defines.keys()
+    if clashes:
+        LOGGER.warning(
+            "Some defines are set both by ESPHome build flags and by LVGL configuration which may lead to unexpected behavior: %s",
+            list(clashes),
+        )
+    unused_defines = all_defines - lv_defines.keys() - defines_from_flags
+
     # Create the content of lv_conf.h with the used defines set to their value, and the unused defines disabled
     definitions = [as_macro(m, v) for m, v in lv_defines.items()] + [
         as_macro(m, "0") for m in unused_defines

--- a/esphome/components/lvgl/__init__.py
+++ b/esphome/components/lvgl/__init__.py
@@ -155,6 +155,8 @@ def generate_lv_conf_h():
     build_flags = (
         CORE.config[CONF_ESPHOME].get(CONF_PLATFORMIO_OPTIONS).get("build_flags", [])
     )
+    if not isinstance(build_flags, list):
+        build_flags = [build_flags]
     # Extract define names from build flags like '-DLV_USE_CHART=1', '-D LV_USE_CHART',
     # or multiple defines in one string.
     define_pattern = r'-D\s*([A-Z_][A-Z0-9_]*)(?:=[^\s\'"\]]*)?'

